ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_prepend(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+	rb_memcached_return(rc);
 }
 
 static VALUE
 rb_connection_append(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(rb_key);
-  char *mvalue = StringValuePtr(rb_value);
-
-  memcached_return_t rc = memcached_append(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(rb_ttl), NUM2INT(rb_flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_append(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+	rb_memcached_return(rc);
 }
 
 static VALUE
 rb_connection_set_prefix(VALUE self, VALUE rb_prefix)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  memcached_return_t rc;
-  if(Qnil == rb_prefix) {
-    rc = memcached_callback_set(ctx->memc, MEMCACHED_CALLBACK_NAMESPACE, NULL);
-  } else {
-    char *prefix = StringValuePtr(rb_prefix);
-    rc = memcached_callback_set(ctx->memc, MEMCACHED_CALLBACK_NAMESPACE, prefix);
-  }
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+	const char *prefix = NULL;
+
+	UnwrapMemcached(self, mc);
+
+	if (!NIL_P(rb_prefix)) {
+		Check_Type(rb_prefix, T_STRING);
+		prefix = StringValueCStr(rb_prefix);
+	}
+
+	rc = memcached_callback_set(mc, MEMCACHED_CALLBACK_NAMESPACE, prefix);
+	rb_memcached_return(rc);
 }
 
 static VALUE
 rb_connection_get_prefix(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  memcached_return_t rc;
-  char * value;
-
-  value = (char*) memcached_callback_get(ctx->memc, MEMCACHED_CALLBACK_NAMESPACE, &rc);
-  handle_memcached_return(rc);
-  if(NULL == value)
-    return Qnil;
-  else
-    return rb_str_new2(value);
-}
+	memcached_st *mc;
+	memcached_return_t rc;
+	const char *value;
 
-static VALUE
-rb_server_to_s(VALUE self)
-{
-  VALUE rb_ret;
-  VALUE rb_hostname = rb_ivar_get(self, id_ivar_hostname);
-  VALUE rb_port = rb_ivar_get(self, id_ivar_port);
-  uint32_t port = NUM2UINT(rb_port);
-
-  rb_ret = rb_str_new("", 0);
-  rb_str_append(rb_ret, rb_hostname);
-  if(0 != port) {
-    rb_str_append(rb_ret, rb_str_new2(":"));
-    rb_str_append(rb_ret, rb_fix2str(rb_port, 10));
-  }
-
-  return rb_ret;
+	UnwrapMemcached(self, mc);
+
+	value = memcached_callback_get(mc, MEMCACHED_CALLBACK_NAMESPACE, &rc);
+	rb_memcached_error_check(rc);
+
+	return value ? rb_str_new2(value) : Qnil;
```

Comment:
This is a null-terminated, proper string, so it's safe to use `rb_str_new2` here.